lib/cache/generic_operations.go

```diff
@@ -138,3 +139,71 @@ func (l genericLister[T, I]) list(ctx context.Context, pageSize int, startToken
 	out, next, err := l.listRange(ctx, pageSize, startToken, "")
 	return out, next, trace.Wrap(err)
 }
+
+// genericRanger is a helper to retrieve a stream from a cache collection.
+type genericRanger[T any, I comparable] struct {
+	// cache to performe the primary read from.
+	cache *Cache
+	// collection that contains the item.
+	collection *collection[T, I]
+	// index of the collection to read with.
+	index I
+	// upstreamRange is the upstream range implementation.
+	upstreamRange func(context.Context, string, string) iter.Seq2[T, error]
+	// fallbackGetter is an optional fallback if upstream does not implment ranging getters.
+	fallbackGetter func(context.Context) ([]T, error)
+}
+
+// Range retrieves a stream of items from the configured cache collection within the range [start, end).
+// If the cache is not healthy, then the items are retrieved from the upstream backend.
+func (g genericRanger[T, I]) Range(ctx context.Context, start, end string) iter.Seq2[T, error] {
+	return func(yield func(T, error) bool) {
+		var zero T
+		rg, err := acquireReadGuard(g.cache, g.collection)
+		if err != nil {
+			yield(zero, err)
```

Comment:
nit: we seem to have standardized on `*new(T)` as our preferred syntax for this kind of thing.

```suggestion
			yield(*new(T), err)
```